compat/optional.h

```diff
@@ -6,8 +6,15 @@
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
 
-#ifndef SQUID__SRC_BASE_OPTIONAL_H
-#define SQUID__SRC_BASE_OPTIONAL_H
+#ifndef SQUID__COMPAT_OPTIONAL_H
+#define SQUID__COMPAT_OPTIONAL_H
+
+#if __cplusplus > 201701L
```

Comment:
If this test survives despite its other, higher-level problems (detailed in other change requests), please use a [standard](https://en.cppreference.com/w/cpp/preprocessor/replace) constant value.

```suggestion
#if __cplusplus >= 201703L
```

[SO](https://stackoverflow.com/a/38456243) has more information about the evolution of these constants. The answer there appears to support the change suggested above.

However, if I missed something, and we have to use a non-standard constant value, then we should document why (via a source code comment).

Reply:
I started with the obvious ```>= 201703L``` test. However testing which OS chose to build the custom code and which used libstd showed some precompilers having trouble with the ```>=``` (operator not supported) resulting in false for the condition.

Also, some libstdc++ libraries defining it to values related to their latest update rather than the standard document release.

Reply:
> However testing which OS chose to build the custom code and which used libstd showed some precompilers having trouble with the >= (operator not supported) resulting in false for the condition.

I fear that those test results were the side effects of other problems, discussed in other change requests. Will wait until those change requests are resolved before asking you to detail your "operator not supported" test results here. For now, please note that Squid uses `>=` in preprocessor code quite a bit already:

```C++
compat/GnuRegex.c:#if (__GNUC__ == 12 && __GNUC_MINOR__ >= 1) || (__GNUC__ > 12)
compat/fdsetsize.h:#if (__GLIBC__ > 2) || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 2)
compat/shm.cc:#if _SQUID_FREEBSD_ && (__FreeBSD__ >= 7)
compat/tempnam.c:#if (L_tmpnam >= L_tmpmin + LONG_BIT / 4)
src/client_side.cc:#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x00908000L
src/ip/Intercept.cc:#if defined(IPFILTER_VERSION) && (IPFILTER_VERSION >= 4000027)
src/security/ServerOptions.cc:#if USE_OPENSSL && OPENSSL_VERSION_NUMBER >= 0x0090800fL && !defined(OPENSSL_NO_ECDH)
src/ssl/support.cc:#if OPENSSL_VERSION_MAJOR >= 3
```